docs/madr/decisions/057-sni-in-the-resource.md

```diff
@@ -0,0 +1,130 @@
+# SNI in the resource definition
+
+* Status: accepted
+
+Technical Story: 
+
+## Context and Problem Statement
+
+SNI is used for cross-zone communication to proxy the traffic through zone proxies.
+Recently we decided on SNI format in MADR 053.
+
+The SNI is constructed on the server and the client side. This approach has a couple of downsides
+* We hit multiple edge cases of constructing it on the client side, because of resource hashing.
+  This logic is not as simple as it should be.
+* Any migration of this format would be hard to execute
+* If we introduce a functionality to enter the mesh directly through zone ingress (assuming that cert is acquired) it would be hard for the client to construct this.
+
+## Considered Options
+
+* SNI in the resource definition
+* Keep it implicit
+
+## Decision Outcome
+
+Chosen option: "Option 1 - SNI in the resource definition", because it solves the problems described in previous section.
+For `MeshMultiZoneService` "Option 2 - compute ports on global cp", because advantages outweights other option.
+
+## Pros and Cons of the Options
+
+### Option 1 - SNI in the resource definition
+
+SNI is unique for each port, therefore we need to define it for each port
+
+```yaml
+kind: MeshService
+spec:
+  ports:
+    - port: 80
+      targetPort: 8080
+      sni: ae10a8071b8a8eeb8.backend.8080.demo.ms # new field
+```
+
+The `sni` field will be filled out automatically be the original Zone CP and synced cross zone.
+The field will be filled out by the same mechanism as Mesh Defaulter, which is webhook on Kubernetes and `ResourceManager` on Universal.
+This way we can avoid a situation that we drop SNI by the resource update.
+
+Advantages:
+* Slightly debuggability because SNI is in MeshService object
+* Easy migration
+* Easy usage from the client side
+
+Disadvantages:
+* Putting more stuff in the object
+
+#### MeshMultiZoneService
+
+SNI should always be filled out by the source of truth of the resource.
+In case of `MeshMultiZoneService`, this is global CP.
+However, the problem is that `MeshMultiZoneService` does not require defining ports directly.
+Global CP does not have list of ports, so it cannot update them with SNI.
+
+```yaml
+kind: MeshMultiZoneService
+name: test-server
+spec:
+  selector:
+    meshService:
+      matchLabels:
+        kuma.io/display-name: test-server
+        k8s.kuma.io/namespace: mzmsconnectivity
+status: # computed on the zone
+  vips:
+    - ip: 241.0.0.1
+  meshServices:
+    - name: test-server.mzmsconnectivity
+    - name: test-server-234jhg34yt34.kuma-system
+  ports:
+    - port: 80
+      targetPort: 8080
+```
+
+##### Option 1 - require ports from user
```

Comment:
What's you suggested option recommendation?

I lean towards 1 just because all the advantages you mention are things that we hit with the implicit `service` we have right now.

So I tend to think if we hit them now, why wouldn't we hit them in a very similar case.

Also it's easier I can imagine people wanting to restrict exposed ports and if we go the future route of auto generating in the future this would not matter.

WDYT?

Reply:
Option 1 is the best for us = engineers because it removes ambiguity.
I worry that it requires more from the user, but maybe it's not that bad?

It's easier to get rid of it in the future than introduce it, so maybe we should start with it

Reply:
I like option 2 more